tests/PS09.md

```diff
@@ -0,0 +1,595 @@
+### PS09xxx - CAN Integration Related Functionality Tests
+
+#### PS09001 - CAN Related Calls - Unauthorized
+
+##### Prerequisites/Inputs
+
+###### Inputs:
+  * Server's IP address
+  * Server's clCustomer port
+
+##### Description 
+
+The test sends *CanStoreDataRequest* and *CanPublishIpnsRecordRequest* to the profile server without checking-in first.
+
+
+###### Step 1:
+
+The test connects to the clCustomer port of the profile server and sends *CanStoreDataRequest*:
+
+  * `Message.id := 1`
+  * `CanStoreDataRequest.data := null`
+ 
+and reads the response. Then it sends *CanPublishIpnsRecordRequest*:
+
+  * `Message.id := 2`
+  * `CanPublishIpnsRecordRequest.record := null`
+
+and reads the response. 
+
+
+
+##### Acceptance Criteria
+
+###### Step 1:
+
+Server replies with *Response*:
+  
+  * `Message.id == 1`
+  * `Response.status == ERROR_UNAUTHORIZED`
+
+Server replies with *Response*:
+  
+  * `Message.id == 2`
+  * `Response.status == ERROR_UNAUTHORIZED`
+
+
+
+
+
+#### PS09002 - CAN Store Data and IPNS Record
+
+##### Prerequisites/Inputs
+  * Server's database is empty.
+  * CAN server is running on localhost
+
+###### Inputs:
+  * Server's IP address
+  * Server's primary port
+  * CAN API port
+
+##### Description 
+
+The test stores object in CAN and sets its IPNS record to point to this object,
+then it refreshes the IPNS record, changes the CAN object and refreshes the IPNS record again.
+Finally, it deletes the CAN object.
+
+
+###### Step 1:
+
+The tests obtains a list of service ports from the profile server's primary port.
+
+The test establishes a hosting agreement with the profile server for its identity and it checks-it in.
+
+
+###### Step 2:
+
+The test sends *CanStoreDataRequest*:
+
+  * `Message.id := 3`
+  * `CanStoreDataRequest.data.hostingServerId` is set to profile server's network ID.
+  * `CanStoreDataRequest.data.keyValueList := 
+     [
+       { key := "key1", stringValue := "value 1" },
+       { key := "key2", uint32Value := 2 },
+       { key := "key3", boolValue := true },
+       { key := "key4", binaryValue := [1,2,3] },
+     ]`
+  * `$IdentityData1 := CanStoreDataRequest.data`
+ 
+and reads the response in form of *CanStoreDataResponse*:
+
+  * `$ObjectHash1 := CanStoreDataResponse.hash`
+
+Then it sends *CanPublishIpnsRecordRequest*:
+
+  * `Message.id := 4`
+  * `$ObjectPath1 := "/ipfs/" + Base58($ObjectHash1)`

```

Comment:
The best thing to do here is probably to explicitly define this in the protocol comments. IPFS is confusing here and I'm not sure if it is a bug or feature. Thing is that those object hashes do have multihash format, but do not have multibase format. So I would expect to see two kinds of prefixes there - one to say it is SHA256 hash and 32 bytes of that and the second saying that it is base58 encoding - and this second is not being used here. Therefore it is probably best to just say what should be there, otherwise it is really confusing (and took me a lot of time to understand while coding actually).